duecredit/io.py

```diff
@@ -72,93 +72,98 @@ def dump(self, tags=None):
                              for k, c in iteritems(citations)
                              if tags.intersection(c.tags))
 
+        packages = {}
+        modules = {}
+        objects = {}
+
+        for key in ('citations', 'entry_keys'):
+            packages[key] = defaultdict(list)
+            modules[key] = defaultdict(list)
+            objects[key] = defaultdict(list)
+
+        # for each path store both a list of entry keys and of citations
+        for (path, entry_key), citation in iteritems(citations):
+            if ':' in path:
+                objects['citations'][path].append(citation)
+                objects['entry_keys'][path].append(entry_key)
+            elif '.' in path:
+                modules['citations'][path].append(citation)
+                modules['entry_keys'][path].append(entry_key)
+            else:
+                packages['citations'][path].append(citation)
+                packages['entry_keys'][path].append(entry_key)
```

Comment:
I would have tried to avoid duplication here
```python
if ':' in path:
    target_dict = objects
elif '.' in path:
    target_dict = modules
else:
    target_dict = packages
target_dict['citations'][path].append(citation)
target_dict['entry_keys'][path].append(entry_key)
```